src/framework/results/experiment_data.hpp

```diff
@@ -177,205 +246,357 @@ void ExperimentData::add_memory_count(const std::string &memory) {
   }
 }
 
-void ExperimentData::add_memory_singleshot(const std::string &memory) {
+void ExperimentData::add_pershot_memory(const std::string &memory) {
   // Memory bits value
   if (return_memory_ && !memory.empty()) {
     memory_.push_back(memory);
   }
 }
 
-void ExperimentData::add_register_singleshot(const std::string &reg) {
+void ExperimentData::add_pershot_register(const std::string &reg) {
   if (return_register_ && !reg.empty()) {
-      register_.push_back(reg);
+    register_.push_back(reg);
   }
 }
 
+//------------------------------------------------------------------
+// Pershot Snapshots
+//------------------------------------------------------------------
 
 template <typename T>
-void ExperimentData::add_singleshot_snapshot(const std::string &type,
-                                          const std::string &label,
-                                          const T &datum) {
-  if (return_snapshots_) {                                              
-    json_t js = datum; // use implicit to_json conversion function for T
-    singleshot_snapshots_[type].add_data(label, js);
+void ExperimentData::add_pershot_snapshot(const std::string &type,
+                                          const std::string &label, T &&datum) {
+  if (return_snapshots_) {
+    // use implicit to_json conversion function for T
+    json_t tmp = datum;
```

Comment:
This function is a bit weird because of perfect forwarding of template param (I couldn't define a `const T&` and a `T&&` version because it's ambiguous). So i'm leaving it up to the template to know if its an lvalue or an rvalue. I can't use `std::move` as it might make what should be a const lvalue into an rvalue.

Reply:
if the compiler is having ambiguous like errors, there must be something else.

Reply:
As we have decided to apply all move semantics related changes to another PR, I'm going to resolve this conversation.